operator/controllers/resources/oidc_connector_controller_test.go

```diff
@@ -0,0 +1,126 @@
+	IssuerURL:    "https://issuer",
+	ClientID:     "client id",
+	ClientSecret: "client secret",
+	ClaimsToRoles: []types.ClaimMapping{{
+		Claim: "claim",
+		Value: "value",
+		Roles: []string{"roleA"},
+	}},
+	RedirectURLs: []string{"https://redirect"},
+}
+
+type oidcTestingPrimitives struct {
+	setup *testSetup
+}
+
+func (g *oidcTestingPrimitives) init(setup *testSetup) {
+	g.setup = setup
+}
+
+func (g *oidcTestingPrimitives) setupTeleportFixtures(ctx context.Context) error {
+	return nil
+}
+
+func (g *oidcTestingPrimitives) createTeleportResource(ctx context.Context, name string) error {
+	oidc, err := types.NewOIDCConnector(name, oidcSpec)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	oidc.SetOrigin(types.OriginKubernetes)
+	return trace.Wrap(g.setup.tClient.UpsertOIDCConnector(ctx, oidc))
+}
+
+func (g *oidcTestingPrimitives) getTeleportResource(ctx context.Context, name string) (types.OIDCConnector, error) {
+	return g.setup.tClient.GetOIDCConnector(ctx, name, true)
+}
+
+func (g *oidcTestingPrimitives) deleteTeleportResource(ctx context.Context, name string) error {
+	return trace.Wrap(g.setup.tClient.DeleteOIDCConnector(ctx, name))
+}
+
+func (g *oidcTestingPrimitives) createKubernetesResource(ctx context.Context, name string) error {
+	oidc := &resourcesv3.TeleportOIDCConnector{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: g.setup.namespace.Name,
+		},
+		Spec: resourcesv3.TeleportOIDCConnectorSpec(oidcSpec),
+	}
+	return trace.Wrap(g.setup.k8sClient.Create(ctx, oidc))
+}
+
+func (g *oidcTestingPrimitives) deleteKubernetesResource(ctx context.Context, name string) error {
+	oidc := &resourcesv3.TeleportOIDCConnector{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: g.setup.namespace.Name,
+		},
+	}
+	return trace.Wrap(g.setup.k8sClient.Delete(ctx, oidc))
+}
+
+func (g *oidcTestingPrimitives) getKubernetesResource(ctx context.Context, name string) (*resourcesv3.TeleportOIDCConnector, error) {
+	oidc := &resourcesv3.TeleportOIDCConnector{}
+	obj := kclient.ObjectKey{
+		Name:      name,
+		Namespace: g.setup.namespace.Name,
+	}
+	err := g.setup.k8sClient.Get(ctx, obj, oidc)
+	return oidc, trace.Wrap(err)
+}
+
+func (g *oidcTestingPrimitives) modifyKubernetesResource(ctx context.Context, name string) error {
+	oidc, err := g.getKubernetesResource(ctx, name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	oidc.Spec.RedirectURLs = []string{"https://redirect1", "https://redirect2"}
+	return g.setup.k8sClient.Update(ctx, oidc)
+}
+
+func (g *oidcTestingPrimitives) compareTeleportAndKubernetesResource(tResource types.OIDCConnector, kResource *resourcesv3.TeleportOIDCConnector) bool {
+	teleportMap, _ := teleportResourceToMap(tResource)
+	kubernetesMap, _ := teleportResourceToMap(kResource.ToTeleport())
+
+	equal := cmp.Equal(teleportMap["spec"], kubernetesMap["spec"])
+	if !equal {
+		fmt.Println(cmp.Diff(teleportMap["spec"], kubernetesMap["spec"]))
```

Comment:
I don't like this, what would be the apropriate way of logging something in the tests?

Reply:
Hmmm, I don't really want to end the tests if this fails, it would end up as a flaky test as this thing is called by `Eventually()`. 

Though, I definitely want to know why this failed, because failure output is `Condition never satisfied`, which does not help to understand why the test did not pass (even with a debugger this is painful to troubleshoot).

Ideally I would have a way to tell `require.Eventually` why each iteration failed, so it can output only the last one. I don't think this is feasible so my plan was to log the diffs: if a test fails, we can look the logs right above and see what the test saw.

Reply:
Instead of using fmt.Println you can use any testing log functions: `t.Log`, `t.Logf` and add a proper logging message instead of just printing the diff between the structures